### ISSUE TITLE
`panic_in_result_fn` shouldn't fire on `todo!()`/`unimplemented!()`/`unreachable!()`

### ISSUE THREAD

**Issue by tgross35** [rust-lang/rust-clippy#11025]
### Summary

`panic_in_result_fn` fires on `panic!`, `unimplemented!`, `todo!`, and `unreachable!`. `panic`,`unimplemented`, and `unreachable` all make since since they represent "final" code. Linting on `todo!` doesn't really make sense in this group however, since it usually represents a work in progress - firing on `todo!()` code is just noisy.

### Lint Name

panic_in_result_fn

### Reproducer

I tried this code:

```rust
#![warn(clippy::panic_in_result_fn)]
fn foo() -> Result<(), ()> { todo!() }
```

I saw this happen:

```
[<output>](warning: used `unimplemented!()`, `unreachable!()`, `todo!()`, `panic!()` or assertion in a function that returns `Result`)
...
```

I expected to see this happen:

No lint

### Version

```text
rustc 1.72.0-nightly (871b59520 2023-05-31)
```

**Comment by charmitro** [comment]
IMO it is a little bit noisy, but I have to ask, is `todo!` any different from `unimplemented!`?

**Comment by tgross35** [comment]
Good point, I was only thinking about my usage of clippy during development. I guess for my use, I probably wouldn’t want  clippy to flag unreachable! or unimemented! either, since both indicate that function shouldn’t be called.

todo! unreachable! and unimplemented! all have non-result use cases and have individual lints, so in my opinion it really only makes sense to fire this on panic!. But then again, it is in restriction…

**Comment by charmitro** [comment]
> Good point, I was only thinking about my usage of clippy during development. I guess for my use, I probably wouldn’t want clippy to flag unreachable! or unimemented! either, since both indicate that function shouldn’t be called.
> 
> todo! unreachable! and unimplemented! all have non-result use cases and have individual lints, so in my opinion it really only makes sense to fire this on panic!. But then again, it is in restriction…

Examining the above macros, we can see that all of them will __always panic__. Since `panic_in_result_fn` job is to check whether a `panic!` is present in functions with `Result`, it seems correct to me.

**Comment by tgross35** [comment]
`unreachable` and `unimplemented` should never actually be reached though in production code, they're used to complete type checking. In other words, reaching them indicates a bug in the program (program is objectively wrong and _should_ fail at runtime) rather than being a part of normal control flow (things you would want to propegate). For example:

```rust
#![allow(unused)]
#![warn(clippy::panic_in_result_fn )]
use std::num::ParseIntError;

#[derive(Debug)]
enum Foo {
    I(u32),
    F(f64),
    S(String)    
}

fn prechecks(foo: &Foo) {
    assert!(!matches!(foo, Foo::F(f64)), "Floats not allowed here!");
}

fn do_things(foo: &Foo) -> Result<u32, ParseIntError> {
    prechecks(foo);
    match foo {
        Foo::I(v) => Ok(*v),
        Foo::S(v) => Ok(v.parse()?)
    }
}
```

This won't compile since the match isn't exhaustive, so it needs an update to:

```rust
    prechecks(foo);
    match foo {
        Foo::I(v) => Ok(*v),
        Foo::S(v) => Ok(v.parse()?),
        Foo::F(_) => unreachable!("verified in prechecks")
    }
```

or

```rust
        Foo::F(_) => unsafe { std::hint::unreachable_unchecked() }
```

This is the only way to correctly complete exhaustiveness checking, and is functionally correct since that variant can't exist at this point. This lint doesn't fire for the second (as expected) but it does fire on the first - I would expect the behavior to match between the two.

They do call `panic` just with specific error messages, but their intended use as type/exhaustivness indicators are pretty different than an `assert` or `panic` that is used to make normal execution infallible (and the [docs](https://doc.rust-lang.org/stable/std/macro.unreachable.html) corraborate that usage)

(end of the day I don't mind turning this lint off, but it just seems a bit unusual that these bug indicators cause it to happen in the first place)

**Comment by Centri3** [comment]
I agree. These shouldn't be linted; `unimplemented` and `todo` represent that this function is not meant to be called, `unreachable` represents that this should never occur.

If somebody wants to, I believe that what it lints is `panic!` in particular but those macros expand to `panic!`. Maybe you can call `root_macro_call`/`root_macro_call_first_node` to get the name of the macro in particular? I dunno.

I believe this could be a good-first-issue. Maybe I'm wrong, anybody feel free to remove it that isn't so ^^

@rustbot label +good-first-issue

**Comment by charmitro** [comment]
@panosfol wanna co-develop this as your first contribution?

**Comment by panosfol** [comment]
> @panosfol wanna co-develop this as your first contribution?

Yeah of course! Thank you!
